File: Examples/iOS/AudioKitDemo/AudioKitDemo/Processing/ConvolutionInstrument.m
//
//  ConvolutionInstrument.m
//  AudioKit Example
//
//  Created by Aurelius Prochazka on 6/27/12.
//  Copyright (c) 2012 Aurelius Prochazka. All rights reserved.
//

#import "ConvolutionInstrument.h"

@implementation ConvolutionInstrument

- (instancetype)initWithInput:(AKAudio *)input
{
    self = [super init];
    if (self) {
        
        // Controls
        _dishWellBalance = [self createPropertyWithValue:0 minimum:0 maximum:1];
        _dryWetBalance   = [self createPropertyWithValue:0 minimum:0 maximum:0.1];
        
        // Instrument definition
        NSString *dish = [AKManager pathToSoundFile:@"dish"      ofType:@"wav"];
        NSString *well = [AKManager pathToSoundFile:@"Stairwell" ofType:@"wav"];
        
        AKConvolution *dishConv = [[AKConvolution alloc] initWithInput:input
                                               impulseResponseFilename:dish];
        
        AKConvolution *wellConv = [[AKConvolution alloc] initWithInput:input
                                               impulseResponseFilename:well];
        
        AKMix *balance = [[AKMix alloc] initWithInput1:dishConv
                                                input2:wellConv
                                               balance:_dishWellBalance];
        
        AKMix *dryWet = [[AKMix alloc] initWithInput1:input
                                               input2:balance
                                              balance:_dryWetBalance];
        
        [self setAudioOutput:[dryWet scaledBy:akp(3.0)]];

        [self resetParameter:input];
        
    }
    return self;
}


@end

File: Examples/iOS/Swift/AudioKitDemo/AudioKitDemo/Processing/ConvolutionInstrument.swift
//
//  ConvolutionInstrument.swift
//  AudioKitDemo
//
//  Created by Nicholas Arner on 3/1/15.
//  Copyright (c) 2015 AudioKit. All rights reserved.
//

class ConvolutionInstrument: AKInstrument
    
{
    // INPUTS AND CONTROLS =====================================================
    
    var dishWellBalance = AKInstrumentProperty(value: 0, minimum: 0, maximum: 1.0)
    var dryWetBalance   = AKInstrumentProperty(value: 0, minimum: 0, maximum: 0.1)
    
    // INSTRUMENT DEFINITION ===================================================
    init(input: AKAudio) {
        super.init()
        
        addProperty(dishWellBalance)
        addProperty(dryWetBalance)
        
        let dish = AKManager.pathToSoundFile("dish", ofType: "wav")
        let well = AKManager.pathToSoundFile("Stairwell", ofType: "wav")
        
        let dishConv = AKConvolution(input: input, impulseResponseFilename: dish!)
        
        let wellConv = AKConvolution(input: input, impulseResponseFilename: well!)
        
        let balance = AKMix(input1: dishConv, input2: wellConv, balance: dishWellBalance)
        
        let dryWet = AKMix(input1: input, input2: balance, balance: dryWetBalance)
        
        // AUDIO OUTPUT ========================================================
        setAudioOutput(dryWet)
        
        // EXTERNAL OUTPUTS ====================================================
        let auxilliaryOutput = AKAudio.globalParameter()
        assignOutput(auxilliaryOutput, to: dryWet)
        
        resetParameter(input)
    }
}